From 54bcd7418f89a1db1099a435ae93e12c857a46a5 Mon Sep 17 00:00:00 2001
From: Vidyullatha Kanchanapally <vidyullatha@codeaurora.org>
Date: Tue, 4 Feb 2014 20:45:38 +0530
Subject: [PATCH] wlan: To remove extra action frame sent in beacon report

Changes have been made in the driver such that the driver
will now not send an extra action frame when there are no
reports on the last channel measured. Instead a frame with
zeroed tagged parameters is only sent when there are no
scan results found on all the channels measured.

Also changes have been made in function
rrmProcessBeaconReportReq to copy measurement duration
and mode correctly when AP channel report is present.

Change-Id: I8a598a98227cf384d4c4178b8ed00aa7d96d9a9e
CRs-Fixed: 594919
---
 CORE/MAC/src/pe/include/rrmGlobal.h |   1 +
 CORE/MAC/src/pe/rrm/rrmApi.c        | 132 +++++++++++++++++++-----------------
 2 files changed, 72 insertions(+), 61 deletions(-)

diff --git a/CORE/MAC/src/pe/include/rrmGlobal.h b/CORE/MAC/src/pe/include/rrmGlobal.h
index de8eda0..c7355f7 100644
--- a/CORE/MAC/src/pe/include/rrmGlobal.h
+++ b/CORE/MAC/src/pe/include/rrmGlobal.h
@@ -184,6 +184,7 @@ typedef struct sRRMReq
          tRRMBeaconReportRequestedIes reqIes;
       }Beacon;
    }request;
+   tANI_U8 sendEmptyBcnRpt;
 }tRRMReq, *tpRRMReq;
 
 typedef struct sRRMCaps
diff --git a/CORE/MAC/src/pe/rrm/rrmApi.c b/CORE/MAC/src/pe/rrm/rrmApi.c
index 23284f6..49c4372 100644
--- a/CORE/MAC/src/pe/rrm/rrmApi.c
+++ b/CORE/MAC/src/pe/rrm/rrmApi.c
@@ -558,6 +558,7 @@ rrmProcessBeaconReportReq( tpAniSirGlobal pMac,
    tANI_U16 measDuration, maxMeasduration;
    tANI_S8  maxDuration;
    tANI_U8  sign;
+   tANI_U16 index;
 
    if( pBeaconReq->measurement_request.Beacon.BeaconReporting.present && 
        (pBeaconReq->measurement_request.Beacon.BeaconReporting.reportingCondition != 0) )
@@ -675,7 +676,8 @@ rrmProcessBeaconReportReq( tpAniSirGlobal pMac,
    pSmeBcnReportReq->channelList.numChannels = num_channels;
    if( pBeaconReq->measurement_request.Beacon.num_APChannelReport )
    {
-      tANI_U8 *pChanList = pSmeBcnReportReq->channelList.channelNumber;
+     tANI_U16 index2 = 0;
+     tANI_U8 *pChanList = pSmeBcnReportReq->channelList.channelNumber;
       for( num_APChanReport = 0 ; num_APChanReport < pBeaconReq->measurement_request.Beacon.num_APChannelReport ; num_APChanReport++ )
       {
          vos_mem_copy(pChanList,
@@ -683,8 +685,11 @@ rrmProcessBeaconReportReq( tpAniSirGlobal pMac,
           pBeaconReq->measurement_request.Beacon.APChannelReport[num_APChanReport].num_channelList);
 
          pChanList += pBeaconReq->measurement_request.Beacon.APChannelReport[num_APChanReport].num_channelList;
-         pSmeBcnReportReq->measurementDuration[num_APChanReport] = SYS_TU_TO_MS(measDuration /*pBeaconReq->measurement_request.Beacon.meas_duration*/);
-         pSmeBcnReportReq->fMeasurementtype[num_APChanReport] = pBeaconReq->measurement_request.Beacon.meas_mode;
+         for( index = 0; index < (pBeaconReq->measurement_request.Beacon.APChannelReport[num_APChanReport].num_channelList); index++ )
+         {
+            pSmeBcnReportReq->measurementDuration[index2] = SYS_TU_TO_MS(measDuration);
+            pSmeBcnReportReq->fMeasurementtype[index2++] = pBeaconReq->measurement_request.Beacon.meas_mode;
+         }
       }
    }
 
@@ -799,8 +804,8 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
                             tpSirBeaconReportXmitInd pBcnReport)
 {
    tSirRetStatus status = eSIR_SUCCESS;
-   tSirMacRadioMeasureReport *pReport;
-   tpRRMReq pCurrentReq = pMac->rrm.rrmPEContext.pCurrentReq; 
+   tSirMacRadioMeasureReport *pReport = NULL;
+   tpRRMReq pCurrentReq = pMac->rrm.rrmPEContext.pCurrentReq;
    tpPESession pSessionEntry ;
    tANI_U8 sessionId;
    v_U8_t flagBSSPresent = FALSE, bssDescCnt = 0;
@@ -817,9 +822,6 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
       return eSIR_FAILURE;
    }
 
-   pBcnReport->numBssDesc = (pBcnReport->numBssDesc == RRM_BCN_RPT_NO_BSS_INFO)?
-                            RRM_BCN_RPT_MIN_RPT : pBcnReport->numBssDesc;
-
    if (NULL == pCurrentReq)
    {
       PELOGE(limLog( pMac, LOGE,
@@ -827,48 +829,54 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
       return eSIR_FAILURE;
    }
 
-   if (NULL == (pSessionEntry = peFindSessionByBssid(pMac,
-                                                     pBcnReport->bssId,
-                                                     &sessionId)))
+   if( (pBcnReport->numBssDesc) ||
+       (!pBcnReport->numBssDesc && pCurrentReq->sendEmptyBcnRpt) )
    {
-      PELOGE(limLog(pMac, LOGE,FL("session does not exist for given bssId"));)
-      return eSIR_FAILURE;
-   }
+      pBcnReport->numBssDesc = (pBcnReport->numBssDesc == RRM_BCN_RPT_NO_BSS_INFO)?
+                               RRM_BCN_RPT_MIN_RPT : pBcnReport->numBssDesc;
+
+      if (NULL == (pSessionEntry = peFindSessionByBssid(pMac,
+                                                        pBcnReport->bssId,
+                                                        &sessionId)))
+      {
+         PELOGE(limLog(pMac, LOGE, FL("session does not exist for given bssId"));)
+         return eSIR_FAILURE;
+      }
 
-   pReport = vos_mem_malloc(pBcnReport->numBssDesc *
-                           sizeof(tSirMacRadioMeasureReport));
+      pReport = vos_mem_malloc(pBcnReport->numBssDesc *
+                              sizeof(tSirMacRadioMeasureReport));
 
-   if (NULL == pReport)
-   {
-      PELOGE(limLog(pMac, LOGE,FL("RRM Report is NULL, allocation failed"));)
-      return eSIR_FAILURE;
-   }
+      if (NULL == pReport)
+      {
+         PELOGE(limLog(pMac, LOGE, FL("RRM Report is NULL, allocation failed"));)
+         return eSIR_FAILURE;
+      }
 
-   vos_mem_zero( pReport,
-                 pBcnReport->numBssDesc * sizeof(tSirMacRadioMeasureReport) );
+      vos_mem_zero( pReport,
+                    pBcnReport->numBssDesc * sizeof(tSirMacRadioMeasureReport) );
 
-   for (bssDescCnt = 0; bssDescCnt < pBcnReport->numBssDesc; bssDescCnt++)
-   {
-       //Prepare the beacon report and send it to the peer.
-       pReport[bssDescCnt].token = pBcnReport->uDialogToken;
-       pReport[bssDescCnt].refused = 0;
-       pReport[bssDescCnt].incapable = 0;
-       pReport[bssDescCnt].type = SIR_MAC_RRM_BEACON_TYPE;
-
-       //If the scan result is NULL then send report request with
-       //option subelement as NULL..
-       if ( NULL != pBcnReport->pBssDescription[bssDescCnt] )
-       {
-           flagBSSPresent = TRUE;
-       }
-
-       //Valid response is included if the size of beacon xmit
-       //is == size of beacon xmit ind + ies
-       if ( pBcnReport->length >= sizeof( tSirBeaconReportXmitInd ) )
-       {
-           pReport[bssDescCnt].report.beaconReport.regClass =  pBcnReport->regClass;
-           if ( flagBSSPresent )
-           {
+      for (bssDescCnt = 0; bssDescCnt < pBcnReport->numBssDesc; bssDescCnt++)
+      {
+         //Prepare the beacon report and send it to the peer.
+         pReport[bssDescCnt].token = pBcnReport->uDialogToken;
+         pReport[bssDescCnt].refused = 0;
+         pReport[bssDescCnt].incapable = 0;
+         pReport[bssDescCnt].type = SIR_MAC_RRM_BEACON_TYPE;
+
+         //If the scan result is NULL then send report request with
+         //option subelement as NULL..
+         if ( NULL != pBcnReport->pBssDescription[bssDescCnt] )
+         {
+            flagBSSPresent = TRUE;
+         }
+
+         //Valid response is included if the size of beacon xmit
+         //is == size of beacon xmit ind + ies
+         if ( pBcnReport->length >= sizeof( tSirBeaconReportXmitInd ) )
+         {
+            pReport[bssDescCnt].report.beaconReport.regClass =  pBcnReport->regClass;
+            if ( flagBSSPresent )
+            {
                pReport[bssDescCnt].report.beaconReport.channel =
                                  pBcnReport->pBssDescription[bssDescCnt]->channelId;
                vos_mem_copy( pReport[bssDescCnt].report.beaconReport.measStartTime,
@@ -890,10 +898,10 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
                vos_mem_copy( pReport[bssDescCnt].report.beaconReport.bssid,
                              pBcnReport->pBssDescription[bssDescCnt]->bssId,
                              sizeof(tSirMacAddr));
-           }
+            }
 
-           switch ( pCurrentReq->request.Beacon.reportingDetail )
-           {
+            switch ( pCurrentReq->request.Beacon.reportingDetail )
+            {
                case BEACON_REPORTING_DETAIL_NO_FF_IE:
                //0 No need to include any elements.
 #if defined WLAN_VOWIFI_DEBUG
@@ -935,20 +943,20 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
                       pBcnReport->pBssDescription[bssDescCnt] );
                }
                break;
-          }
-       }
-   }
+            }
+         }
+      }
 
-#if defined WLAN_VOWIFI_DEBUG
-   PELOGE(limLog( pMac, LOGE, "Sending Action frame ");)
-#endif
-   limSendRadioMeasureReportActionFrame( pMac,
-                                         pCurrentReq->dialog_token,
-                                         bssDescCnt,
-                                         pReport,
-                                         pBcnReport->bssId,
-                                         pSessionEntry );
+      limLog( pMac, LOG1, "Sending Action frame with %d bss info", bssDescCnt);
+      limSendRadioMeasureReportActionFrame( pMac,
+                                            pCurrentReq->dialog_token,
+                                            bssDescCnt,
+                                            pReport,
+                                            pBcnReport->bssId,
+                                            pSessionEntry );
 
+      pCurrentReq->sendEmptyBcnRpt = false;
+   }
 
    if( pBcnReport->fMeasureDone )
    {
@@ -957,7 +965,8 @@ rrmProcessBeaconReportXmit( tpAniSirGlobal pMac,
       rrmCleanup(pMac);
    }
 
-   vos_mem_free(pReport);
+   if( NULL != pReport )
+      vos_mem_free(pReport);
 
    return status;
 }
@@ -1130,6 +1139,7 @@ rrmProcessRadioMeasurementRequest( tpAniSirGlobal pMac,
                 vos_mem_set(pCurrentReq, sizeof( *pCurrentReq ), 0);
                pCurrentReq->dialog_token = pRRMReq->DialogToken.token;
                pCurrentReq->token = pRRMReq->MeasurementRequest[i].measurement_token;
+               pCurrentReq->sendEmptyBcnRpt = true;
                pMac->rrm.rrmPEContext.pCurrentReq = pCurrentReq;
                rrmStatus = rrmProcessBeaconReportReq( pMac, pCurrentReq, &pRRMReq->MeasurementRequest[i], pSessionEntry );
                if (eRRM_SUCCESS != rrmStatus)
-- 
1.8.2.1

